File: solutions/fortran/rna-transcription/1/rna_transcription.f90
module rna_transcription
  implicit none
contains

  function to_rna(dna)
    character(*) :: dna
    character(len(dna)) :: to_rna
    integer :: i
    do i = 1, len(dna)
       to_rna(i:i) = translate(dna(i:i))
    end do
  end function to_rna

  function translate(dna)
    character :: dna, translate
    select case (dna)
    case ('G')
       translate = 'C'
    case ('C')
       translate = 'G'
    case ('T')
       translate = 'A'
    case ('A')
       translate = 'U'
    case default
       translate = ' '
    end select
  end function translate
end module rna_transcription

File: solutions/c/rna-transcription/1/rna_transcription.c
#define _POSIX_C_SOURCE 1
#include "rna_transcription.h"
#include <string.h>

char *to_rna(const char *dna) {
  char *rna = strdup(dna);
  for (char *c = rna; *c; c++) {
    if (*c == 'A') {
      *c = 'U';
    } else if (*c == 'C') {
      *c = 'G';
    } else if (*c == 'G') {
      *c = 'C';
    } else if (*c == 'T') {
      *c = 'A';
    }
  }
  return rna;
}
